File: Linked_List/2487.Remove_Nodes_From_Linked_List/Python/2487.Remove_Nodes_From_Linked_List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNodes(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        #==============================#
        # Stack-based traversal method #
        #==============================#

        ############
        #Initialize
        ##### Record stack #####
        record_stack = []

        ############################
        #Stack-based loop traversal
        while (head):
            while (record_stack and ((record_stack[(-1)]).val < head.val)):
                record_stack.pop() #Keep updating, popped

            ##### Check if the current stack matched conditions or not #####
            if (record_stack):
                (record_stack[(-1)]).next = head #Keep updating/recording
            else:
                pass

            record_stack.append(head) #Keep updating/recording

            head = head.next #Keep updating/traversing

        return record_stack[0]

File: Linked_List/2487.Remove_Nodes_From_Linked_List/C++/2487.Remove_Nodes_From_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode * removeNodes(ListNode * head) 
    {
        //========================================//
        // Input type:                            //
        //   - ListNode * (head list-node)        //
        // Return type:                           //
        //   - ListNode * (result head list-node) //
        //========================================//

        /*Initialize*/
        ///// Record vector /////
        vector<ListNode * > record_vect;

        /*Stack-based loop traversal*/
        while (head != nullptr) //Whole
        {
            while ((! record_vect.empty()) && (((record_vect.back())->val) < (head->val))) //Part
            {
                record_vect.pop_back(); //Keep updating, popped

            } //Part

            ///// Check if the current vector matched conditions or not /////
            if (! record_vect.empty()) { ((record_vect.back())->next) = head; } //Keep updating, recording
            else { ; }

            record_vect.emplace_back(head); //Keep updating, recording

            head = (head->next); //Keep updating, traversing

        } //Whole

        return record_vect.front();
    }
};
